### PR TITLE
Change RecordBatch::schema return type from SchemaRef to &SchemaRef

### PR DESCRIPTION
**Pull request by monkwire**
# Which issue does this PR close?
This addresses the first part of #5342.

# Rationale for this change
Prior to this change, `RecordBatch::schema` returned a `SchemaRef`, which in most cases can only be used as an owned value.  By instead returning `&SchemaRef`, users will have more flexibility in the way they use `RecordBatch::schema`; it's return value can be used as a borrowed value or cloned into an owned value.

# What changes are included in this PR?
`RecordBatch::schema` returns `&SchemaRef` instead of `SchemaRef`. I made minor changes to  parts of the code where `RecordBatch::schema` is called.

# Are there any user-facing changes?
Yes, this will affect the way that users use `RecordBatch::schema`.

### CODE REVIEW

**Comment by Monkwire3** [comment]
I believe this is a breaking change, but I'm not sure how to add the appropriate label.

File: arrow-flight/tests/flight_sql_client_cli.rs [hunk 1]
```diff
@@ -245,7 +245,7 @@ impl FlightSqlService for FlightSqlServiceImpl {
             "part_2" => batch.slice(2, 1),
             ticket => panic!("Invalid ticket: {ticket:?}"),
         };
-        let schema = batch.schema();
+        let schema = batch.schema().clone();
         let batches = vec![batch];
         let flight_data = batches_to_flight_data(schema.as_ref(), batches)
```
Comment (tustvold) [inline arrow-flight/tests/flight_sql_client_cli.rs:250]:
```suggestion
        let schema = batch.schema();
        let batches = vec![batch];
        let flight_data = batches_to_flight_data(schema.as_ref(), batches)
```

File: arrow-flight/examples/flight_sql_server.rs [hunk 2]
```diff
@@ -643,7 +643,8 @@ impl FlightSqlService for FlightSqlServiceImpl {
         self.check_token(&request)?;
         let schema = Self::fake_result()
             .map_err(|e| status!("Error getting result schema", e))?
-            .schema();
+            .schema()
+            .clone();
         let message = SchemaAsIpc::new(&schema, &IpcWriteOptions::default())
```
Comment (tustvold) [inline arrow-flight/examples/flight_sql_server.rs:648]: I think we could avoid this clone

**Review by tustvold — APPROVED** [review]
This looks good to me.

I suspect this will be quite disruptive downstream so tagging @alamb

File: arrow-flight/tests/flight_sql_client_cli.rs (hunk 1 above)
Comment (monkwire) [reply arrow-flight/tests/flight_sql_client_cli.rs:250]: I tried this change on my machine, but I got a borrow checker error. How does this look instead?
```suggestion
        let schema = batch.schema();
        let batches = vec![batch.clone()];
        let flight_data = batches_to_flight_data(schema, batches)
```

File: arrow-flight/examples/flight_sql_server.rs (hunk 2 above)
Comment (monkwire) [reply arrow-flight/examples/flight_sql_server.rs:648]: Thanks, I was able to remove this `.clone()` by saving the `RecordBatch` to a non-temporary value.

**Review by alamb — COMMENTED** [review]
Thank you for this contribution @monkwire 

I think the change to return `&SchemaRef` (and forcing callsites to call `clone()`) is the more rust ideomatic pattern and if there was no existing users of arrow-rs would be the right API

However, I think this change will cause massive code churn in all users of arrow-rs for very little practical gain (it will permit saving `Arc::clone` in some places). I believe the magnitude of the impact on consumers of the library is likely huge, estimated by the number of places that are required in this PR.

Thus I am not in favor of this change, and I apologize I did not comment on the PR sooner 

I reread https://github.com/apache/arrow-rs/issues/5342 and I have an alternate suggestion on https://github.com/apache/arrow-rs/issues/5342#issuecomment-1975825386 which I think would be much less disruptful

**Comment by tustvold** [comment]
Perhaps we could add a `schema_ref` method in that case, and keep the `schema` method unchanged?

**Comment by alamb** [comment]
> Perhaps we could add a `schema_ref` method in that case, and keep the `schema` method unchanged?

This would make sense to me.

**Comment by alamb** [comment]
https://github.com/apache/arrow-rs/issues/5463 for the larger discussion

**Comment by tustvold** [comment]
Closing in favour of #5474

**Pull request closed without merging**